src/category_theory/groupoid/subgroupoid.lean

```diff
@@ -0,0 +1,471 @@
+    { rintros,
+      simp only [Inter_coe_set, mem_Inter] at hp hq ⊢,
+      exact λ S Ss, S.mul' (hp S Ss) (hq S Ss), } } ⟩
+
+instance : complete_lattice (subgroupoid C) :=
+{ bot          := (⊥),
+  bot_le       := λ S, empty_subset _,
+  top          := (⊤),
+  le_top       := λ S, subset_univ _,
+  inf          := (⊓),
+  le_inf       := λ R S T RS RT ⟨c,d,f⟩ h, ⟨RS h, RT h⟩,
+  inf_le_left  := λ R S ⟨c,d,f⟩ pRS, pRS.left,
+  inf_le_right := λ R S ⟨c,d,f⟩ pRS, pRS.right,
+  .. complete_lattice_of_Inf (subgroupoid C)
+       ( by
+        { dsimp only [Inf], rintro s, constructor,
+          { rintro S Ss ⟨c,d,f⟩,
+            simp only [Inter_coe_set, mem_Inter, mem_coe_set_iff],
+            exact λ hp, hp S Ss, },
+          { rintro T Tl ⟨c,d,f⟩ fT,
+            simp only [Inter_coe_set, mem_Inter, mem_coe_set_iff],
+            exact λ S Ss, (Tl Ss) fT, }}) }
+
+lemma le_objs {S T : subgroupoid C} (h : S ≤ T) : S.objs ⊆ T.objs :=
+λ s ⟨γ, hγ⟩, ⟨γ, by { rw ←mem_coe_set_iff at hγ ⊢, exact h hγ, }⟩
+
+/-- The functor associated to the embedding of subgroupoids -/
+def inclusion {S T : subgroupoid C} (h : S ≤ T) : S.coe ⥤ T.coe :=
+{ obj := λ s, ⟨s.val, le_objs h s.prop⟩,
+  map := λ s t f, ⟨f.val, by { rw ←mem_coe_set_iff, apply h, rw mem_coe_set_iff, exact f.prop, } ⟩,
+  map_id' := λ _, rfl,
+  map_comp' := λ _ _ _ _ _, rfl }
+
+lemma inclusion_inj_on_objects {S T : subgroupoid C} (h : S ≤ T) :
+  function.injective (inclusion h).obj :=
+begin
+  rintros ⟨s,hs⟩ ⟨t,ht⟩ he,
+  simp only [inclusion, subtype.mk_eq_mk] at he ⊢,
+  exact he,
+end
+
+lemma inclusion_faithful {S T : subgroupoid C} (h : S ≤ T) (s t : S.coe):
+  function.injective (λ (f : s ⟶ t), (inclusion h).map f) :=
+begin
+  dsimp only [inclusion],
+  rintros ⟨f,hf⟩ ⟨g,hg⟩ he,
+  simp only [subtype.mk_eq_mk] at he ⊢,
+  exact he,
+end
+
+lemma inclusion_refl {S : subgroupoid C} : inclusion (le_refl S) = 𝟭 S.coe :=
+begin
+  dsimp only [inclusion],
+  fapply functor.ext,
+  { rintros,
+    simp only [subtype.val_eq_coe, subtype.coe_eta, functor.id_obj], },
+  { rintros ⟨s,hs⟩ ⟨t,ht⟩ ⟨f,hf⟩,
+    simp only [eq_to_hom_refl, functor.id_map, category.comp_id, category.id_comp,
+               subtype.mk_eq_mk], }
+end
+
+lemma inclusion_trans {R S T : subgroupoid C} (k : R ≤ S) (h : S ≤ T) :
+  inclusion (k.trans h) = (inclusion k) ⋙ (inclusion h) :=
+begin
+  dsimp only [inclusion],
+  fapply functor.ext,
+  { rintros, simp only [functor.comp_obj], },
+  { rintros ⟨s,hs⟩ ⟨t,ht⟩ ⟨f,hf⟩,
+    simp only [eq_to_hom_refl, functor.comp_map, category.comp_id, category.id_comp,
+    subtype.mk_eq_mk], }
+end
+
+lemma inclusion_comp_embedding {S T : subgroupoid C} (h : S ≤ T) :
+  (inclusion h) ⋙ T.hom = S.hom :=
+begin
+  dsimp only [inclusion, hom],
+  fapply functor.ext,
+  { rintros, simp only [functor.comp_obj, subtype.val_eq_coe], },
+  { rintros ⟨s,hs⟩ ⟨t,ht⟩ ⟨f,hf⟩,
+    simp only [functor.comp_map, subtype.coe_mk, eq_to_hom_refl, category.comp_id,
+    category.id_comp], }
+end
```

Comment:
These two are actually defeq; with definitional eta in Lean 4, `inclusion_refl` probably is also a defeq.